### PR TITLE
Add integration tests for kube-apiextensions-server

### PR DESCRIPTION
**Pull request by nikhita**
**What this PR does / why we need it**: Add the following integration tests for `kube-apiextensions-server`:
- [x] test namespace scoped resources
- [x] test cluster scoped resources
- [x] test discovery
- [x] test no namespace rejects
- [x] test same name, different namespace

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: for #45511 

**Special notes for your reviewer**:

**Release note**:

```
NONE
```

@sttts @deads2k

### CODE REVIEW

**Comment by sttts** [comment]
@k8s-bot ok to test

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 1]
```diff
@@ -186,3 +188,204 @@ func TestSimpleCRUD(t *testing.T) {
 		t.Errorf("missing watch event")
 	}
 }
+
+func TestNonNamespaced(t *testing.T) {
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: I guess there is quite some overlap with TestSimpleCRUD. Can you try to factor out the equal pieces?

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/testserver/resources.go [hunk 2]
```diff
@@ -28,19 +28,38 @@ import (
...
+				Singular:   "nonenglishnoxu",
+				Kind:       "WishIHadChosenNoxu",
+				ShortNames: []string{"shortnoxu", "shortnoxu2"},
+				ListKind:   "NoxuItemList",
+			},
+		},
+	}
+}
+
+func NewNoxuClusterCustomResourceDefinition() *apiextensionsv1alpha1.CustomResource {
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/testserver/resources.go]: Also here I would merge these two funcs and add `scope` parameter to `NewNoxuCustomResourceDefinition`.

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 3]
```diff
@@ -76,14 +88,14 @@ func TestSimpleCRUD(t *testing.T) {
...
+	noxuWatch, err := noxuResourceClient.Watch(metav1.ListOptions{ResourceVersion: initialListListMeta.GetResourceVersion()})
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer noxuNamespacedWatch.Stop()
+	defer noxuWatch.Stop()
 
 	noxuInstanceToCreate := testserver.NewNoxuInstance(ns, "foo")
-	createdNoxuInstance, err := noxuNamespacedResourceClient.Create(noxuInstanceToCreate)
+	createdNoxuInstance, err := noxuResourceClient.Create(noxuInstanceToCreate)
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: A check is missing below whether ObjectMeta.Namespace is set correctly.

Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: Also check that resources with the wrong scope are rejected.

And check that creating another resource with another namespace, but same name works or fails (depending on scope).

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 4]
```diff
@@ -186,3 +198,49 @@ func TestSimpleCRUD(t *testing.T) {
...
+	if r.Name != "noxus" {
+		t.Fatalf("Expected exactly the resource \"noxus\" in group version %v/%v via discovery, got: %v", group, version, r.Name)
+	}
+	if r.Kind != "WishIHadChosenNoxu" {
+		t.Fatalf("Expected exactly the kind \"WishIHadChosenNoxu\" in group version %v/%v via discovery, got: %v", group, version, r.Kind)
+	}
+
+	s := []string{"shortnoxu", "shortnoxu2"}
+	if !reflect.DeepEqual(r.ShortNames, s) {
+		t.Fatalf("Expected exactly the shortnames \"shortnoxu\", \"shortnoxu2\" in group version %v/%v via discovery, got: %v", group, version, r.ShortNames)
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: Nit:

```
`foo bar "abc" def`
```

is a bit more elegant than quoting.

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 5]
```diff
@@ -186,3 +198,49 @@ func TestSimpleCRUD(t *testing.T) {
 		t.Errorf("missing watch event")
 	}
 }
+func TestDiscovery(t *testing.T) {
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: empty line

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 6]
```diff
@@ -167,3 +182,97 @@ func TestSimpleCRUD(t *testing.T) {
...
+	createdNoxuInstance, err := instantiateCustomResource(t, testserver.NewNoxuInstance(ns, "foo"), noxuResourceClient, noxuDefinition)
+	if err == nil {
+		t.Fatalf("unexpected non-error while creating noxu instance: %v", createdNoxuInstance)
+	}
+}
+
+func TestSameNameDiffNamespace(t *testing.T) {
+	scope := apiextensionsv1alpha1.NamespaceScoped
+	ns1 := "namespace-1"
+	testSimpleCRUD(t, scope, ns1)
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: each of these creates its own server, doesn't it? We have to share it though for this test.

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 7]
```diff
@@ -167,3 +182,97 @@ func TestSimpleCRUD(t *testing.T) {
...
+	}
+	if e, a := noxuDefinition.Spec.Group+"/"+noxuDefinition.Spec.Version, initialListTypeMeta.GetAPIVersion(); e != a {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+	if e, a := noxuDefinition.Spec.Names.ListKind, initialListTypeMeta.GetKind(); e != a {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+
+	createdNoxuInstance, err := instantiateCustomResource(t, testserver.NewNoxuInstance(ns, "foo"), noxuResourceClient, noxuDefinition)
+	if err == nil {
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: which error do we expect?

File: staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go [hunk 8]
```diff
@@ -168,14 +184,359 @@ func TestSimpleCRUD(t *testing.T) {
...
+		if e, a := createdObjectMeta1.GetUID(), deletedObjectMeta1.GetUID(); e != a {
+			t.Errorf("expected %v, got %v", e, a)
+		}
+
+	case <-time.After(5 * time.Second):
+		t.Errorf("missing watch event")
+	}
+
+	// Second namespace
+	ns2 := "namespace-2"
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/test/integration/basic_test.go]: Is this much different from namespace-1? Can we merge these two into a (local) func?

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Pull request merged**